blog article #03: Visualizing Nginx Access Logs using Timeplus and Grafana

ayewo opened the pull request:
PR checklist:
- Did you run ClangFormat ? N/A
- Did you separate headers to a different section in existing community code base ? N/A
- Did you surround `proton: starts/ends` for new code in existing community code base ? N?A

Please write user-readable short description of the changes:

ayewo committed:
Intro for article 03.

ayewo committed:
add Introduction

ayewo committed:
formatting

ayewo committed:
WIP

ayewo committed:
Docker and related files

ayewo committed:
Flesh out section

jovezhong commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,79 @@
+  
+The first two require access to an AWS account. The last two products need to have been running in production, in order to have meaningful access log data to experiment with. 
+
+This article does away with most of those prerequisities making it easier to get started. It builds on the same analysis of the previous article but you only need: 
```
prerequisities is a tpo. 

For "This article does away with most of those prerequisities", I don't quite get it. Do you need you still need those prerequisites to run the examples in the blog, or doesn't need it. Maybe change a way to express it.

ayewo replied:
Thanks for catching the typo.

I've reworded a bit. Hopefully it's clearer.

jovezhong commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,79 @@
+* a Grafana container;
+* a third container that seeds the Timeplus Proton container with access log data.
+
+The 3rd container uses the `proton client` to execute an SQL file [`01_nginx_access_log.sql`](01_nginx_access_log.sql) used to seed Timeplus Proton with access log. That SQL file uses the [`CREATE RANDOM STREAM`](https://docs.timeplus.com/proton-create-stream#create-random-stream) DDL to create a stream named `nginx_access_log`. 
```
I would suggest showing the main DDL of the random stream, don't have to show all settings, but give the reader an idea of how the random stream is defined without looking into the sql file (also when we publish the blog on our website, the sql file will remain in github proton repo. I guess many ppl will be too lazy to check the link. For important information, please put in the blog content directly)

ayewo replied:
I decided to include a screenshot of the DDL so I could do a line-by-line explanation of each database function used to generate random data **and** also link out to the relevant docs.

jovezhong commented on examples/nginx-grafana-metabase/README.md:
````diff
@@ -0,0 +1,79 @@
+1 row in set. Elapsed: 0.013 sec. 
+```
+
+Due to this design decision, we will need to first take a snapshot from the random stream. This snapshot will represent our access log data so we can reliably geolocate the latitude and longitude of each (randomly generated) IP address before we can correctly visualize them in Grafana.
````
you can use a materialized view to save the random stream data from the query to an inner stream.

ayewo replied:
I tried different `MATERIALIZED VIEW` incantations but each time I queried the materialized view, it acted exactly like the underlying `RANDOM STREAM` it was backing.

I got a different random record each I time ran the query:
```sql
SELECT remote_ip, rfc1413_ident, remote_user, date_time, 
http_verb, path, http_ver, status, size, referer, user_agent, malicious_request 
FROM nginx_access_log_view LIMIT 1;
```

jovezhong reviewed (commented):
Good start. I reckon this PR is working-in-progress. Please add more content with screenshots. Look forwards to the final PR

ayewo committed:
code screenshot with line numbers and update others

ayewo committed:
add explanation for DDL generation

ayewo committed:
finish table of random stream defaults

ayewo committed:
explain the structure of nginx_access_log

ayewo committed:
reword explanation

ayewo committed:
formatting & reword

ayewo committed:
done with nginx

ayewo committed:
rewordingg

ayewo committed:
access log data generation is working

ayewo committed:
minor rewording

jovezhong commented on examples/nginx-grafana-metabase/README.md:
````diff
@@ -0,0 +1,162 @@
+}
+```
+
+Below is a brief explanation of how each access log line is parsed into 12 columns:
````
I would suggest explain the log_format first, such as $request is a string separated by whitespace.

Then you share your design how to turn this plain text format to a table/stream with 12 columns.

Otherwise the reader will have some challenges to immediately get why there are 12 columns.

jovezhong commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,162 @@
+* `-` is stored in column `rfc1413_ident` of type `string`;
+* `$remote_user` is stored in column `remote_user` of type `string`;
+* `$time_local` os stored in column `date_time` of type `datetime64`;
+* `$request` is split into 4 columns each of type `string` to make our analysis easier:
```
better to mention this string is separated by whitespace.  I would suggest omitting `;` in the column column, the user may get the wrong impression each part is separated by ; not white space. Also I suggest we talk about 4 parts in the requests, then mention how they are mapped to columns in Timeplus stream schema

jovezhong commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,162 @@
+* `$http_referer` is stored in column `referer` of type `string`;
+* `$http_user_agent` is stored in column `user_agent` of type `string`.
+
+The previous article created a stream with a total of 12 columns to take advantage of Timeplus Proton's fast aggregation for our analysis. We will reuse the same approach in this article. 
```
Now that you talk about column, maybe have a table in Markdown to show such 12 columns, with column name, data type, meaning and sample value

jovezhong commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,162 @@
+Just like in the previous article, the stream also has 12 columns. Each column is randomly assigned a `default` value as you can see on lines 4 - 30.
+
+Using the line numbers 4 - 30 as reference, the table below goes over each column and the database [functions](https://docs.timeplus.com/functions) used to generate test data.
+| Line  | Column | `default` | 
```
yes, the table is nice. I guess there might be a better way to organize the content. Maybe have this table right after you explain the log format. Also the user won't really compare the image with the table, so having the line number column is not very helpful.

jovezhong commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,162 @@
+Due to this design decision, we will need to create a snapshot from the random stream to serve as our test data. This snapshot will represent our access log data so we can reliably geolocate the latitude and longitude of each (randomly generated) IP address before we can correctly visualize them in Grafana.
+
+Afterwards, the article will use the random stream to illustrate streaming queries in Grafana.
+
```
Look forwards to more content about the SQL and Grafana integation.

ayewo replied:
Yes, of course.

I get occasional stability issues when using the Timeplus plugin with Grafana that requires me to restart the Docker container, e.g.:
```bash
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:42.119403134Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:46.375972748Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=892 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:46.376686483Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:47.136723585Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:47.137064412Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=900 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:51.393905513Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=907 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:51.393964919Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:52.146066063Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=915 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:52.146175035Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:56.411043054Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=924 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:56.411227803Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:57.167190451Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=931 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:57.167402103Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:01.42450791Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=939 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:01.42481641Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:02.181165871Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:02.181467678Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=947 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:06.440446079Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=955 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:06.440538286Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:07.201749037Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=963 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:07.201899997Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=grafana-apiserver t=2024-07-14T11:19:07.73712243Z level=info msg="[core] [Channel #245 SubChannel #246]grpc: addrConn.createTransport failed to connect to {Addr: \"unused\", ServerName: \"unused\", }. Err: connection error: desc = \"transport: error while dialing: dial unix /tmp/plugin3391027809: connect: connection refused\""
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:11.454859501Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=971 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:11.454996798Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:12.217220448Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=979 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:12.217427932Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
```

It seems the stability issues come from when I try to use the plugin for streaming queries.
After a fresh restart of the containers, everything is fine but once I add a streaming query to a dashboard in Grafana, the plugin process eventually dies:
```bash
03_nginx-grafana-metabase-grafana-1     | logger=sqlstore t=2024-07-14T11:27:30.9359683Z level=info msg="Created default admin" user=admin
03_nginx-grafana-metabase-grafana-1     | logger=sqlstore t=2024-07-14T11:27:30.93639324Z level=info msg="Created default organization"
03_nginx-grafana-metabase-grafana-1     | logger=secrets t=2024-07-14T11:27:30.939589903Z level=info msg="Envelope encryption state" enabled=true currentprovider=secretKey.v1
03_nginx-grafana-metabase-grafana-1     | logger=plugin.angulardetectorsprovider.dynamic t=2024-07-14T11:27:30.976876557Z level=info msg="Restored cache from database" duration=375.537µs
03_nginx-grafana-metabase-grafana-1     | logger=plugin.store t=2024-07-14T11:27:30.978859549Z level=info msg="Loading plugins..."
03_nginx-grafana-metabase-grafana-1     | logger=plugins.registration t=2024-07-14T11:27:31.007469113Z level=error msg="Could not register plugin" pluginId=xychart error="plugin xychart is already registered"
03_nginx-grafana-metabase-grafana-1     | logger=plugins.initialization t=2024-07-14T11:27:31.007554961Z level=error msg="Could not initialize plugin" pluginId=xychart error="plugin xychart is already registered"
03_nginx-grafana-metabase-grafana-1     | logger=local.finder t=2024-07-14T11:27:31.007645178Z level=warn msg="Skipping finding plugins as directory does not exist" path=/usr/share/grafana/plugins-bundled
03_nginx-grafana-metabase-grafana-1     | logger=plugins.registration t=2024-07-14T11:27:31.56854336Z level=info msg="Plugin registered" pluginId=timeplus-proton-datasource
03_nginx-grafana-metabase-grafana-1     | logger=plugin.store t=2024-07-14T11:27:31.568596056Z level=info msg="Plugins loaded" count=55 duration=589.737958ms
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.dashboard t=2024-07-14T11:27:31.57261121Z level=error msg="can't read dashboard provisioning files from directory" path=/etc/grafana/provisioning/dashboards error="open /etc/grafana/provisioning/dashboards: no such file or directory"
03_nginx-grafana-metabase-grafana-1     | logger=query_data t=2024-07-14T11:27:31.572757561Z level=info msg="Query Service initialization"
03_nginx-grafana-metabase-grafana-1     | logger=live.push_http t=2024-07-14T11:27:31.57509053Z level=info msg="Live Push Gateway initialization"
03_nginx-grafana-metabase-grafana-1     | logger=ngalert.notifier.alertmanager org=1 t=2024-07-14T11:27:31.578696115Z level=info msg="Applying new configuration to Alertmanager" configHash=d2c56faca6af2a5772ff4253222f7386
03_nginx-grafana-metabase-grafana-1     | logger=ngalert.state.manager t=2024-07-14T11:27:31.584592127Z level=info msg="Running in alternative execution of Error/NoData mode"
03_nginx-grafana-metabase-grafana-1     | logger=infra.usagestats.collector t=2024-07-14T11:27:31.58638178Z level=info msg="registering usage stat providers" usageStatsProvidersLen=2
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.datasources t=2024-07-14T11:27:31.586668827Z level=error msg="can't read datasource provisioning files from directory" path=/etc/grafana/provisioning/datasources error="open /etc/grafana/provisioning/datasources: no such file or directory"
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.plugins t=2024-07-14T11:27:31.587240916Z level=error msg="Failed to read plugin provisioning files from directory" path=/etc/grafana/provisioning/plugins error="open /etc/grafana/provisioning/plugins: no such file or directory"
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.alerting t=2024-07-14T11:27:31.58738615Z level=error msg="can't read alerting provisioning files from directory" path=/etc/grafana/provisioning/alerting error="open /etc/grafana/provisioning/alerting: no such file or directory"
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.alerting t=2024-07-14T11:27:31.587420396Z level=info msg="starting to provision alerting"
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.alerting t=2024-07-14T11:27:31.587431968Z level=info msg="finished to provision alerting"
03_nginx-grafana-metabase-grafana-1     | logger=grafanaStorageLogger t=2024-07-14T11:27:31.587693419Z level=info msg="Storage starting"
03_nginx-grafana-metabase-grafana-1     | logger=ngalert.state.manager t=2024-07-14T11:27:31.588656011Z level=info msg="Warming state cache for startup"
03_nginx-grafana-metabase-grafana-1     | logger=ngalert.multiorg.alertmanager t=2024-07-14T11:27:31.588813805Z level=info msg="Starting MultiOrg Alertmanager"
03_nginx-grafana-metabase-grafana-1     | logger=http.server t=2024-07-14T11:27:31.596650277Z level=info msg="HTTP Server Listen" address=[::]:3000 protocol=http subUrl= socket=
03_nginx-grafana-metabase-grafana-1     | logger=sqlstore.transactions t=2024-07-14T11:27:31.607687482Z level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=0 code="database is locked"
03_nginx-grafana-metabase-grafana-1     | logger=ngalert.state.manager t=2024-07-14T11:27:31.63795999Z level=info msg="State cache has been initialized" states=0 duration=49.297312ms
03_nginx-grafana-metabase-grafana-1     | logger=ngalert.scheduler t=2024-07-14T11:27:31.638049334Z level=info msg="Starting scheduler" tickInterval=10s maxAttempts=1
03_nginx-grafana-metabase-grafana-1     | logger=ticker t=2024-07-14T11:27:31.638142031Z level=info msg=starting first_tick=2024-07-14T11:27:40Z
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.dashboard t=2024-07-14T11:27:31.639171522Z level=info msg="starting to provision dashboards"
03_nginx-grafana-metabase-grafana-1     | logger=provisioning.dashboard t=2024-07-14T11:27:31.639275153Z level=info msg="finished to provision dashboards"
03_nginx-grafana-metabase-grafana-1     | logger=grafana-apiserver t=2024-07-14T11:27:31.89963796Z level=info msg="Adding GroupVersion playlist.grafana.app v0alpha1 to ResourceManager"
03_nginx-grafana-metabase-grafana-1     | logger=grafana-apiserver t=2024-07-14T11:27:31.900688298Z level=info msg="Adding GroupVersion featuretoggle.grafana.app v0alpha1 to ResourceManager"
03_nginx-grafana-metabase-grafana-1     | logger=grafana.update.checker t=2024-07-14T11:27:32.364555258Z level=info msg="Update check succeeded" duration=773.798305ms
03_nginx-grafana-metabase-grafana-1     | logger=plugin.angulardetectorsprovider.dynamic t=2024-07-14T11:27:32.390862109Z level=info msg="Patterns update finished" duration=743.410427ms
03_nginx-grafana-metabase-grafana-1     | logger=plugins.update.checker t=2024-07-14T11:27:32.45243534Z level=info msg="Update check succeeded" duration=861.036091ms
03_nginx-grafana-metabase-grafana-1     | logger=context userId=0 orgId=1 uname= t=2024-07-14T11:27:45.842166713Z level=info msg="Request Completed" method=GET path=/dashboard status=404 remote_addr=172.18.0.1 time_ms=15 duration=15.935553ms size=44906 referer= handler=notfound status_source=server
03_nginx-grafana-metabase-grafana-1     | logger=context userId=0 orgId=1 uname= t=2024-07-14T11:27:58.021398478Z level=info msg="Request Completed" method=GET path=/api/live/ws status=-1 remote_addr=172.18.0.1 time_ms=3 duration=3.451001ms size=0 referer= handler=/api/live/ws status_source=server
03_nginx-grafana-metabase-grafana-1     | logger=infra.usagestats t=2024-07-14T11:28:52.600521021Z level=info msg="Usage stats are ready to report"
03_nginx-grafana-metabase-grafana-1     | logger=live t=2024-07-14T11:31:27.904714728Z level=info msg="Initialized channel handler" channel=ds/edrq3l2naqe4gb/f9157113-0ea4-4ce7-b6fe-13817e342079 address=ds/edrq3l2naqe4gb/f9157113-0ea4-4ce7-b6fe-13817e342079

------>>>> process DIES here:
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:31:37.89472788Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=32 error="exit status 2"
------>>>> BUT Grafana fails to restart the process afterwards:

03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:37.894978435Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=0s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:37.896452629Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="[plugin.unavailable] plugin unavailable" wait=0s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:37.896658734Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="[plugin.unavailable] plugin unavailable" wait=0s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:38.099283201Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="[plugin.unavailable] plugin unavailable" wait=200ms
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:38.501169814Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="[plugin.unavailable] plugin unavailable" wait=400ms
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:31:39.322271281Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=53 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:39.322407808Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=800ms
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:31:40.935940391Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=60 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:40.935977112Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=1.6s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:31:44.150730736Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=67 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:44.150952105Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=3.2s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:49.159487291Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:31:49.159776643Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=76 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:31:54.179511992Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=85 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:31:54.179795905Z level=error msg="Error running stream, re-establishing" path=f9157113-0ea4-4ce7-b6fe-13817e342079 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
```

jovezhong reviewed (changes requested):
I shared some comments. Please make the PR as draft PR, as it's not complete yet.  I am on travel these 2 weeks so may not provide timely review.  Once the PR is ready to review, I will spend more time on this.

ayewo committed:
focus exclusively on Grafana

ayewo committed:
focus on grafana in blog 03

ayewo committed:
small refactoring

ayewo committed:
1st line example

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
update regex to match the image

ayewo committed:
reword

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
wip

ayewo committed:
replace

ayewo committed:
wip 2

ayewo committed:
wip 3

ayewo committed:
wip 3

ayewo committed:
wip 4

ayewo committed:
replace image

ayewo committed:
replace image again

ayewo committed:
replace img

ayewo committed:
remove

ayewo committed:
Add files via upload

ayewo committed:
add updated files

ayewo committed:
step-by-step

ayewo committed:
minor

ayewo committed:
add screenshots

ayewo committed:
rename file names

ayewo committed:
add links to screenshots

ayewo commented:
Hi @jovezhong!

I've addressed all of your review comments. This PR is ready for review so hopefully you'll be able to take a look soon.

I hope you are enjoying your travels ✈️ :)

jovezhong commented:
Thank you. I will share my feedback in 48 hours, if not sooner.

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,340 @@
+Before I dive into the article proper, I will first revisit the format used by Nginx to record access log data so that the data we'll generate will mimic its shape and properties.
+
+## The Shape of Nginx's Access Log Data
+In the [previous article](https://www.timeplus.com/post/analyzing-nginx-access-logs), we briefly touched on the Nginx access logs but did not dwell on its location or how the data it contains should be parsed.
```
"its location". I am a bit surprised to see "location". Do you mean how this is generated or data is parsed?

ayewo replied:
Location in this context meant where the Nginx access logs is stored on disk. I've reworded it.

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,340 @@
+
+The first line in the access log is from the IP address `161.35.230.x`. 
+
+![nginx access log - first line](images/01_nginx-access-log-1st-line.png)
```
Thanks for the nice screenshots. I guess you may use tools like https://carbon.sh/ to generate it with background. However since your blog will be posted on timeplus.com with dark background. It will be great if you can use that dark color `#0F0D10` to replace the blue background. You may have to change the text color too. Sorry for the trouble

ayewo replied:
I hear you. 
Yes, I used Carbon and thought that the background I picked would work well with the T+ blog.

I used the Preview.app on macOS which meant creating each one took a lot of time. Maybe I'll find time to recreate them but it shouldn't prevent you from approving this, if there are no other blockers.

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,340 @@
+
+In fact, this was why we had a total of 12 columns in the `nginx_historical_access_log` stream created in the previous article.
+
+Below is a summary of how individual fields from an Nginx access log will map to database columns:
```
Better to change "database columns" to "Timeplus columns"

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,340 @@
+|1 | `remote_ip` | [`random_in_type('ipv4')`](https://docs.timeplus.com/functions_for_random#random_in_type): returns a random [ipv4](https://docs.timeplus.com/datatypes) representing a user's IP address. |
+| 2 | `rfc1413_ident` | `default`s to '-' but should be a string conforming to [RFC1413](https://datatracker.ietf.org/doc/html/rfc1413). Currently unused. |
+| 3 | `remote_user` | `default`s to '-' since most blog traffic is from unauthenticated users. Currently unused. |
+| 4 | `date_time` | `random_in_type('datetime64', 365, y -> to_time('2023-6-17') + interval y day)`: [`random_in_type('datetime64')`](https://docs.timeplus.com/functions_for_random#random_in_type) between 2023-6-17 & a 1-yr interval i.e. [2023-06-17, 2024-06-17). |
```
very smart!

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,340 @@
+
+### Random Streams
+There is an important note about the design of random streams in Timeplus Proton mentioned in the [`RANDOM STREAM`](https://docs.timeplus.com/proton-create-stream#create-random-stream) documentation that is worth highlighting here:
+> [!NOTE]
```
I realized there could be a misunderstanding based on my previous doc, that people may think all data generated will be kept in the memory. I changed that doc to the following text. Please quote it or highlight it in your own words

When you run a Timeplus SQL query with a random stream, the data will be generated and analayzed by the query engine. Depending on the query, all generated data or aggregated states can be kept in memory during the query time. If you are not querying the random stream, there is no data generated or kept in memory.

ayewo replied:
It's still there. Perhaps you updated a different page?

On https://docs.timeplus.com/proton-create-stream#create-random-stream when I do a Ctrl+F for this text: 
> The data of random stream is kept in memory during the query time. If you are not querying the random stream, there is no data generated or kept in memory.

jovezhong replied:
I will publish that doc change in 1 or 2 hours. Same link

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,340 @@
+11. Now choose the Timeplus Proton data source you setup earlier:
+![](images/11.jpeg)
+
+12. Congrats! You've successfully set up the "Nginx Access Logs Analysis" dashboard in Grafana:
```
The steps are quite long. Thanks for the screenshots for each step. This certainly help but maybe you can just pre-configure the docker compose to setup the dashboard and source.  You can check my example at https://github.com/timeplus-io/proton/tree/develop/examples/grafana You may mount it at ./grafana_provisioning:/etc/grafana/provisioning In the folder it can define the Timeplus proton data source and use that source in the dashboard JSON.

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,340 @@
+![](images/11.jpeg)
+
+12. Congrats! You've successfully set up the "Nginx Access Logs Analysis" dashboard in Grafana:
+![](images/12.jpeg)
```
As a reader, I am very excited to see this big dashboard. The  blog ends right here, which is a bit disappointing. 

Maybe you can add a few screenshot to explain each panel to highlight why it's helpful to understand the data pattern.

You can also show the SQL behind.

If you can use streaming SQL, then maybe this can be a live dashboard to show the current traffic. In that case, a GIF will be very impressive

ayewo replied:
Added the SQL along with screen shots, as requested.

I had some difficulty with the streaming functionality as I commented here https://github.com/timeplus-io/proton/pull/798#discussion_r1677113108 so I decided to leave it out so I could ship this post as it was taking too long already ... :)

jovezhong reviewed (changes requested):
Again @ayewo , this is a very nice blog with huge potential. Not just for ppl who wants to learn more about streaming processing, for network engineers or anyone who wants to learn more about nginx, network traffic monitoring, or SQL for O11y, this will be a great hit.

I shared my comments. It will great to change the background of the code snippet from blue to dark color as our blog. But this can be time consuming and you can do this later or don't have to do it.

As a reader, I am very excited to see this big dashboard. But the blog ends right here, which is a bit disappointing. 

Maybe you can add a few screenshot to explain each panel to highlight why it's helpful to understand the data pattern.

You can also show the SQL behind.

If you can use streaming SQL, then maybe this can be a live dashboard to show the current traffic. In that case, a GIF will be very impressive.

Please also add a short section to summarize the blog and highlight some takeaways.

ayewo committed:
update container name

ayewo committed:
automatic provision of dashboard

ayewo committed:
replace image

ayewo committed:
text edits

heyitswei commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,9 @@
+# Visualizing Nginx Access Logs using Grafana and Metabase
+In my last [blog post](https://www.timeplus.com/post/analyzing-nginx-access-logs), I showed how Timeplus Proton can be used to analyze web traffic in production using just the log files generated by an Nginx web server. The article explored real-time analysis and historical analysis but the analysis wasn't visualized. 
```
Hi Saïd, thank you for writing this article! Here are some minor wording edits: 

... web traffic in production, using just the log files... 

- [ ] Add a comma after "in production"

ayewo replied:
Done.

heyitswei commented on examples/nginx-grafana-metabase/README.md:
```diff
@@ -0,0 +1,9 @@
+# Visualizing Nginx Access Logs using Grafana and Metabase
+In my last [blog post](https://www.timeplus.com/post/analyzing-nginx-access-logs), I showed how Timeplus Proton can be used to analyze web traffic in production using just the log files generated by an Nginx web server. The article explored real-time analysis and historical analysis but the analysis wasn't visualized. 
+
+This article will remedy the lack of rich visualizations by showing how to do so using open source tools like [Grafana](https://grafana.com/) and [Metabase](https://www.metabase.com/).
```
Can make this sentence more concise: 

This article remedies the lack of rich visualizations by showing how to use open-source tools like Grafana.

ayewo replied:
Fixed.

heyitswei commented on examples/nginx-grafana/README.md:
```diff
@@ -246,8 +244,85 @@ Query id: 11d071e4-6028-47da-826d-9db42176f8e5
+5. Open http://localhost:3000 in your web browser:
+
+
+6. Click on the hamburger menu at the top left and click on "Data sources" to navigate to the "Data sources" page so you can setup a new data source:
```
Reduce wordiness, and change "setup" to "set up"

Click on the hamburger menu at the top left and click on "Data sources" to set up a new data source on that page:

ayewo replied:
It's no longer there. Seems like you are reviewing an old version of this file?

ayewo committed:
add summary

ayewo committed:
add grafana panel images

ayewo committed:
add SQL code behind all 5 Grafana panels

ayewo committed:
fix images

ayewo commented:
@jovezhong 

All feedback addressed!

jovezhong commented on examples/nginx-grafana/README.md:
```diff
@@ -0,0 +1,381 @@
+# Visualizing Nginx Access Logs using Grafana
```
Maybe change the title to "Visualizing Nginx Access Logs using Timeplus and Grafana", with a better context when we post and share the blog outside timeplus.com

jovezhong reviewed (approved):
Thanks for the nice blog. Approved (better to change the title to add Timeplus)

jovezhong commented:
/tip 225

ayewo committed:
Update title
